include/mapnik/feature_style_processor.hpp

```diff
@@ -113,7 +115,24 @@ class MAPNIK_DECL feature_style_processor
     void render_material(layer_rendering_material const & mat, Processor & p );
 
     Map const& m_;
+
+#ifdef MAPNIK_METRICS
+public:
+    void enable_metrics(bool status);
+    std::shared_ptr<metrics> get_metrics();
+
+private:
+    bool metrics_enabled_ = false;
+    std::shared_ptr<metrics> metrics_ = nullptr;
```

Comment:
Why `std::shared_ptr`? I think this line should be 

```c++
metrics & metrics_;
```

Lifetime of `metrics` object should be responsibility of an observer - caller of Mapnik. Except of semantics, `std::shared_ptr` is complicated thing with some overhead.

Reply:
Or `metrics * metrics_;`, if it should be nullable.

Reply:
Basically, it's a shared pointer to ease working with a metrics object in different threads without needing an strict owner.
Let's say you have 1 thread handling a query that, internally creates 5 tasks (different threads also), 2 of them being non-cancelable. If you want to cancel the query while it's running and the main thread is the only responsible (`metrics *`) then it has to wait for all other threads to stop to free it, if it holds a shared_ptr then it can remove its reference, return immediately and the memory will be freed when the last task stops.

I know right now it's possible to just go for raw pointer but if I have to choose between some overhead and the automatic memory management I'd go for the later.

Reply:
OK, it might be right place for shared ownership. Overhead was not my primary concern.

Why is the same code duplicated in `grid.hpp`, `image.hpp`, `image_any.hpp` and  `image_null.hpp`?

Reply:
I didn't want to use inheritance for such different classes with some of them already inheriting, but I could create a virtual class with the basic members (metrics and metrics enabled) and getter/setters if you prefer it.

Reply:
What about compositing?

Reply:
Yeah, that could be a good option. I'll have a look and see if I can set up another class (or modify metrics) to make it public so everything can be accessed directly just by declaring a new member; e.g: `grid.metrics.enable()` or `this.metrics.time("metric.name")`